File: unit_tests/test_unary.cpp
#include "compiler.hpp"
#include "simulator.hpp"
#include "exceptions.hpp"

namespace tests {
	TEST_F(CompilerTest, TestBitwise) {
		std::string source = R"(
int main() {
	return ~12;
})";
		compile(source, ss);
		std::cout << ss.str() << std::endl;
		simulator.loadProgram(ss.str());
		EXPECT_EQ(~12, static_cast<int>(simulator.execute()));
	}

	TEST_F(CompilerTest, TestBitwise0) {
		std::string source = R"(
int main() {
	return ~0;
})";
		compile(source, ss);
		simulator.loadProgram(ss.str());
		EXPECT_EQ(~0, static_cast<int>(simulator.execute()));
	}

	TEST_F(CompilerTest, TestMissingConst) {
		std::string source = R"(
int main() {
	return ~;
})";
		EXPECT_THROW(compile(source, ss), syntax_error);
	}

	TEST_F(CompilerTest, TestMissingSemicolon2) {
		std::string source = R"(
int main() {
	return ~12
})";
		EXPECT_THROW(compile(source, ss), syntax_error);
	}

	TEST_F(CompilerTest, TestWrongOrder) {
		std::string source = R"(
int main() {
	return 12~;
})";
		EXPECT_THROW(compile(source, ss), syntax_error);
	}
}

File: tests/test_unary.rs
// tests/test_unary.rs
mod simulator;

use rstest::*;
use simulator::{CompilerTest, harness};
use compiler::CompilerError;

#[rstest]
fn test_bitwise(mut harness: CompilerTest) {
    let source = r#"
int main() {
    return ~12;
}
"#;
    harness.assert_runs_ok(source, !12);
}

#[rstest]
fn test_bitwise0(mut harness: CompilerTest) {
    let source = r#"
int main() {
    return ~0;
}
"#;
    harness.assert_runs_ok(source, !0);
}

#[rstest]
fn test_missing_const(harness: CompilerTest) {
    let source = r#"
int main() {
    return ~;
}
"#;
    assert_compile_err!(harness, source, CompilerError::SyntaxError(_));
}

#[rstest]
fn test_missing_semicolon2(harness: CompilerTest) {
    let source = r#"
int main() {
    return ~12
}
"#;
    assert_compile_err!(harness, source, CompilerError::SyntaxError(_));
}

#[rstest]
fn test_wrong_order(harness: CompilerTest) {
    let source = r#"
int main() {
    return 12~;
}
"#;
    assert_compile_err!(harness, source, CompilerError::SyntaxError(_));
}